### ISSUE TITLE
trigger_rule=TriggerRule.ONE_FAILED doesn't work properly with task_groups

### ISSUE THREAD

**Issue by antonio-antuan** [apache/airflow#30333]
### Apache Airflow version

2.5.2 (checked on 2.5.3 also)

### What happened

I'd like to setup "watcher" pattern inside of task_group, but the task always marked as "skipped".
reference to a similar issue: #30334 30334

### How to reproduce

```
from datetime import datetime

from airflow import DAG
from airflow.decorators import task_group
from airflow.operators.python import task
from airflow.utils.trigger_rule import TriggerRule


@task(retries=0, trigger_rule=TriggerRule.ONE_FAILED)
def watcher():
    pass


@task
def foo(x):
    pass

@task
def failed(x):
    raise Exception("Failed")


@task_group()
def process_attestation(x):
    [foo(x), failed(x)] >> watcher()


with DAG(
        dag_id='test',
        start_date=datetime.now(),
) as dag:
    # Receives sboms from database, one row per each DAG run.
    # Will be changed in the future with HTTP-endpoint or something like that.
    @task
    def provider():
        return [1, 2, 3]


    process_attestation.expand(x=provider())

```

### Operating System

Arch Linux (kernel 6.2.6)

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==7.3.0
apache-airflow-providers-common-sql==1.3.4
apache-airflow-providers-ftp==3.3.1
apache-airflow-providers-google==8.11.0
apache-airflow-providers-http==4.2.0
apache-airflow-providers-imap==3.1.1
apache-airflow-providers-postgres==5.4.0
apache-airflow-providers-sqlite==3.3.1

### Deployment

Docker-Compose

### Anything else

the result graph screenshot:
![image](https://user-images.githubusercontent.com/6426720/227979509-58344733-cb97-46b6-bd57-50dc11b1eab1.png)

task instance details screenshot:
![image](https://user-images.githubusercontent.com/6426720/227979307-3622bb76-6f50-4b6b-9776-d6aa257d9642.png)

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by antonio-antuan** [comment]
checked on 2.5.3, still reproduced

**Comment by pankajastro** [comment]
look like this happening when trying to expand on task_group.

**Comment by hussein-awala** [comment]
@pankajastro yes, this happens only with mapped task groups, but I confirm that there is a bug on this part. Unfortunately I didn't have time to work on it.

**Comment by eladkal** [comment]
@dzhigimont are you still working on this issue?

**Comment by dzhigimont** [comment]
Hi @eladkal I am working on it. I had some pauses in the process

**Comment by uranusjr** [comment]
OK, so I went on a debug adventure, and here’s my conclusion. The analysis in https://github.com/apache/airflow/pull/32701#issuecomment-1647371922 is spot on, but the problem is deeper. Currently, we first check whether a task has all its dependencies satisfied, and only expand it if so:

https://github.com/apache/airflow/blob/9f3af9c862faec73cd477e8398b7607d2d045b37/airflow/models/dagrun.py#L853-L869

(Notice how `are_dependencies_met` must satisfy to reach `_expand_mapped_task_if_needed`)

Whether a trigger rule dependency is satisfied, however, is not known before expansion, since we only want to consider “relevant” upstreams. This is why #32802 doesn’t work. I actually don’t think #32701 completely fixes the issue either—if I understand the code correctly, it will fail when we add nested task mapping (i.e. a mapped task inside a mapped task group). But I guess that’s a problem we can address when (if) we ever reach there.

I’ll slightly tweak #32701 to enhance the test case and add some comments for later viewers and approve it.

**Comment by uranusjr** [comment]
On further consideration, I don’t think #32701 should be merged as-is. While marking all upstream tis as relevant fixed ONE_FAILED, it would create a bug in the reverse direction for things such as ALL_FAILED, since in that case, TriggerRuleDep would over-consider upstream tis.

I think TriggerRuleDep needs a much more significant rewrite to implement a proper fix.

**Comment by red-crown** [comment]
Is there any kind of workaround for this issue?

**Comment by shahar1** [comment]
The DAG in the `How to reproduce` section works as expected when running it using the current `main` branch (v2.9.2+), so I think it is safe to presume it has been fixed since v2.5.2. However, I cannot put my finger on the exact PR that fixed it (maybe #33903 ?) - feel free to comment if you're certain about that part.
Closing as resolved.
